documentation/src/main/asciidoc/user_guide/clustering.adoc

```diff
@@ -258,111 +355,96 @@ the initial nodes do not appear within the specified timeout, the cache manager
 ==== Server Hinting
-The motivations behind this feature is to ensure when using distribution, backups are not
-picked to reside on the same physical server, rack or data centre.
-For obvious reasons it doesn't work with total replication.
+The following topology hints can be specified:
+
+Machine::
+  This is probably the most useful, when multiple JVM instances run on the
+  same node, or even when multiple virtual machines run on the same physical machine.
+Rack::
+  In larger clusters, nodes located on the same rack are more likely to experience a
+  hardware or network failure at the same time.
+Site::
+  Some clusters may have nodes in multiple physical locations for extra resilience.
+  Note that <<CrossSiteReplication, Cross site replication>> is another alternative for
+  clusters that need to span two or more data centres.
+
+All of the above are optional.
+When provided, the distribution algorithm will try to spread the ownership of each
+segment across as many sites, racks, and machines (in this order) as possible.
 
 ===== Configuration
 The hints are configured at transport level:
 
 [source,xml]
 ----
 <transport
-    cluster = "MyCluster"
-    machine = "LinuxServer01"
-    rack = "Rack01"
-    site = "US-WestCoast" />
+    cluster="MyCluster"
+    machine="LinuxServer01"
+    rack="Rack01"
+    site="US-WestCoast" />
 ----
 
-The following topology hints can be specified:
-
-machine::
-  This is probably the most useful, to disambiguate between multiple JVM instances on the
-  same node, or even multiple virtual hosts on the same physical host.
-rack::
-  In larger clusters with nodes occupying more than a single rack, this setting would help
-  prevent backups being stored on the same rack.
-site::
-  To differentiate between nodes in different data centres replicating to each other. Note
-  that <<CrossSiteReplication, Cross site replication>> is another alternative for
-  clusters that need to span two or more data centres.
-
-All of the above are optional, and if not provided, the distribution algorithms provide no
-guarantees that backups will not be stored in instances on the same machine/rack/site.
-
 
 [[KeyAffinityService]]
 ====  Key affinity service
-The key affinity service solves the following problem: for a distributed Infinispan
-cluster one wants to make sure that a value is placed in a certain node.
-Based on a supplied cluster
-link:http://docs.jboss.org/infinispan/{infinispanversion}/apidocs/org/infinispan/remoting/transport/Address.html[address]
-identifying the node, the service returns a key that will be hashed to that particular
-node.
+In a distributed cache, a key is allocated to a list of nodes with an opaque algorithm.
+There is no easy way to reverse the computation and generate a key that maps to a
+particular node.
+However, we can generate a sequence of (pseudo-)random keys, see what their primary
+owner is, and hand them out to the application when it needs a key mapping to a
+particular node.
```

Comment:
The `AffinityPartitioner` works more like grouping, because it can only guarantee that 2 keys are mapped to the same owners and not that a key is mapped to a particular node.